components/sync15/src/sync_multiple.rs

```diff
@@ -51,45 +53,37 @@ impl ClientInfo {
 /// succeeded.
 pub fn sync_multiple(
     stores: &[&dyn Store],
-    persisted_global_state: &Cell<Option<String>>,
-    last_client_info: &Cell<Option<ClientInfo>>,
+    persisted_global_state: &mut Option<String>,
+    mem_cached_state: &mut MemoryCachedState,
     storage_init: &Sync15StorageClientInit,
     root_sync_key: &KeyBundle,
     sync_ping: &mut telemetry::SyncTelemetryPing,
 ) -> result::Result<HashMap<String, Error>, Error> {
-    // Note: We explicitly swap a None back as the state, meaning if we
-    // unexpectedly fail below, the next sync will redownload meta/global,
-    // crypto/keys, etc. without needing to. Apparently this is both okay
-    // and by design.
-    let persisted = persisted_global_state.replace(None);
-    let maybe_global = match persisted {
+    let mut pgs = match persisted_global_state {
         Some(persisted_string) => {
-            match serde_json::from_str::<GlobalState>(&persisted_string) {
-                Ok(state) => Some(state),
+            match serde_json::from_str::<PersistedGlobalState>(&persisted_string) {
+                Ok(state) => state,
                 _ => {
                     // Don't log the error since it might contain sensitive
-                    // info like keys (the JSON does, after all).
-                    log::error!("Failed to parse GlobalState from JSON! Falling back to default");
-                    None
+                    // info (although currently it only contains the declined engines list)
+                    log::error!(
+                        "Failed to parse PersistedGlobalState from JSON! Falling back to default"
+                    );
+                    PersistedGlobalState::default()
                 }
             }
         }
-        None => None,
-    };
-
-    let mut global_state = match maybe_global {
-        Some(g) => g,
         None => {
-            log::info!("First time through since unlock. Creating default global state.");
-            last_client_info.replace(None);
-            GlobalState::default()
+            log::warn!("The application didn't give us persisted state - this is only expected on the very first run");
+            PersistedGlobalState::default()
         }
     };
 
-    // Ditto for the ClientInfo - if we fail below the cell will have None, so
-    // will be re-initialized on the next sync.
-    let client_info = match last_client_info.replace(None) {
+    // We put None back into last_client_info now so if we fail entirely,
+    // reinitialize everything related to the client.
+    let client_info = match mem::replace(&mut mem_cached_state.last_client_info, None) {
```

Comment:
As mentioned, moving from Cell::replace => RefCell that gets mem::replaced is not really entirely a step forward. But hopefully accurately modeling stuff using &mut wont be too bad.

Reply:
I've killed these RefCells